README.md
[Auto-Sort Bookmarks](https://addons.mozilla.org/en-US/firefox/addon/auto-sort-bookmarks/)
==========================================================================================

Firefox Plugin to Sort Bookmarks by Multiple Criteria

Description
-----------

Provide a way to sort all bookmarks and automatically sort bookmarks.

This extension considers separators, so that it sorts bookmarks between separators.
This extension adds an icon to sort all bookmarks to the toolbar. To remove it, you must choose the Customize option  in the context menu on the Firefox user interface.

**Please backup your current bookmarks in case you do not like the new bookmarks order. Thus, you could restore them.**

This extension provides a few options categorized within these categories:
* Main Options
* Exclude Folders
* Sort Criteria
* Sort Order

These options work like this:

**Main Options:**

* **Auto-sort :** if this option is enabled, the bookmarks will be sorted when Firefox is opened, when this option is activated and when bookmarks are added, changed, moved or deleted.
This means you cannot move any bookmarks in the same folder, unless it is moved over a separator.
* **Delay:** allow to define a delay (in seconds) before automatically sorting bookmarks.
* **Delay for Folders:** allow to define a delay (in seconds) before automatically sorting folders. This is to avoid that a new folder is sorted before you can choose it when adding a new bookmark.
* **Case Insensitive:** if activated, the bookmarks will be sorted without considering the letter case.

**Exclude Folders:**

This button opens a new tab allowing you to exclude folders when sorting. If you uncheck the checkbox next to a folder, it wont be sorted, but the children folders will be sorted.
If you want to exclude a folder recursively from being sorted, check the recursive checkbox.

**Sort Criteria:**

* **Sort By:** allow to specify the first sort criteria, that is to say, the order that will be used to sort the bookmarks. The choices are : name, url, description, keyword, date added, last modified, last visited, visited count and reversed base-URL.
* **Inverse Order:** if this option is enabled, the order specified in "Sort By" will be reversed. So the order will be descending.
* **Then Sort By:** allow to specify a second sort criteria (optional). For instance, if the first sort criteria is the name, it is possible to choose a second sort criteria to sort bookmarks with the same name.
* **Inverse Second Order:** if this option is enabled, the order specified in "Then Sort By" will be reversed.
* **Sort Folder By:** allow to specify a different sort criteria for folders. For instance, you might want to sort folders by name and other kinds of bookmarks by last visited.
* **Inverse Folder Order:** if this option is enabled, the order speficied in "Sort Folder By" will be reversed.

**Sort Order:**

* **Folder Sort Order :** this option specify the order in which the folders are sorted. For instance, it is possible to sort folders before livemarks, smart bookmarks and bookmarks. To do this, the folder sort order should be lesser than the other sort orders.
If the sort order of two types are at the same level, the bookmarks from these types will be sorted together.
* **Livemark Sort Order:** this option specify the sort order of livemarks.
* **Smart Bookmark Sort Order:** this option specify the sort order of smart bookmarks
* **Bookmark Sort Order:** this option specify the bookmark sort order.


History.md
2.10 / 2015-08-22
===================

 * Fixed issue and improved performance in folders to sort page.

2.9 / 2015-08-13
==================

 * Added missing german translation.
 * Fixed issue causing hidden text in folders to sort page.

2.8 / 2015-08-06
==================

 * Removed menu items and added action button.
 * Added option for case insensitive sort.
 * Added option to exclude folders when sorting.
 * Added option for folder sort order.
 * Fixed bookmarks not sorting when there are corrupted bookmarks.
 * Added new sort by reversed base url.

2.7 / 2015-01-14
==================

 * Fixed for Firefox 36.

2.5 / 2014-10-30
==================

 * Fixed icon and menuitem position in bookmarks menu.
 * Fixed bad preferences persistence.
 * Fixed bookmarks not sorting in some cases.

2.4 / 2014-10-17
==================

 * Added germain translations.
 * Removed style in add-on option page.
 * Improved bookmark sorter to support numeric sorting (1, 2, 3, …, 10 instead of 1, 10, 2, 3, …).
 * Replaced panel shown at first install by a tab.
 * Fixed issues about auto-sorting not working.

2.3 / 2013-12-12
==================

 * Fixed compatibility with Firefox 26.

2.1 / 2013-08-06
==================

 * Added labels to categorize the preference page.
 * Changed close icon of the confirmation dialog.
 * Removed the icon in the add-on bar in prevision of the new Firefox interface.
 * Moved the confirmation panel to the star button.
 * Fixed error when trying to add a sort bookmarks item in a hidden menu.
 * Fixed memory leaks.
 * Fixed menu item not added when adding the bookmarks toolbar button.
 * Removed the SDK from the package.
 * Changed the compatible versions in install.rdf.

2.0.2 / 2013-07-17
====================

 * Complete rewrite with the Add-on SDK.
 * Now restartless.
 * Compatibility fixes.
 * Added options to disallow the sort in the menu and the unsorted folders.
 * Added a confirmation after install before enabling the auto sorting.
 * Added delay and folder delay options (fixes the bug stopping the user to add a bookmark in a new folder).
 * Added two sort criteria options (with the possibility to reverse the sorting).
 * Added the menu item to sort all bookmarks in all bookmarks menus and the organise menu of the bookmarks manager (with options to hide them).

1.6 / 2012-11-20
==================

 * Readd menu item and toolbar icon that dissapeared in version 1.3.

1.3 / 2012-10-14
==================

 * Add the option to set the sort order.
 * Fix bug that didn't sort bookmarks when removing a separator.
 * Fix bug that slows down the browser when sorting too much bookmarks.
 * Fix bug that adds bookmark observer more than once.

1.2 / 2012-09-28
==================

1.1 / 2012-09-26
==================

1.0 / 2012-08-28
==================


lib/main.js
/*
 * Copyright (C) 2014-2015  Boucher, Antoni <bouanto@zoho.com>
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

'use strict';

/**
 *
 * TODO: add tests for new features (including testing with corrupted bookmarks, folders, separators, ...).
 * TODO: fix test-main, test-bookmark-sorter.
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 * TODO: use the Places Query API with the provided sort option.
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 * TODO: add style rule to disallow space before comma.
 * TODO: add style rule to disallow writing function* ().
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 * TODO: whenever an option is changed, call sortIfAuto().
 * TODO: translate the options in French.
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 * TODO: only delay the folder sort for new folder when the user is in the new bookmark dialog.
 *
 * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *
 *
 * FIXME: bug when moving a separator or moving across a separtor in auto sort mode (cannot reproduce — use the bookmark observer to log the actions you tried).
 * TODO: look for errors on rare case (sorting and deleting the same folder or trying to move a deleted item — make some error tests).
 *
 */

const _ = require('sdk/l10n').get;
const { has, reset, set } = require('sdk/preferences/service');
const self = require('sdk/self');
const data = self.data;
const simplePrefs = require('sdk/simple-prefs');
const prefs = simplePrefs.prefs;
const tabs = require('sdk/tabs');
const { ActionButton } = require('sdk/ui/button/action');
const windowUtils = require('sdk/window/utils');
const { setDoNotSortAnnotation, setRecursiveAnnotation } = require('annotations');
const { BookmarkManager, Folder, Separator, menuFolder, toolbarFolder, unsortedFolder } = require('bookmarks');
const bookmarkManager = new BookmarkManager({});
const { BookmarkSorter } = require('bookmark-sorter');
const { showConfigureFoldersToExclude } = require('configure-folders');
const { getOptionName, setPreferenceMaximum, setPreferenceMinimum } = require('options');
const SECOND = 1000;
const sortCriterias = [
        'title',
        'url',
        'description',
        'keyword',
        'dateAdded',
        'lastModified',
        'lastVisited',
        'accessCount',
        'revurl',
    ];

let bookmarkSorter = new BookmarkSorter();

/**
 * On begin batch callback.
 */
function onBeginBatch() {
    bookmarkSorter.stop();
}

/**
 * Sort `item` parent folder after the preferred delay.
 * @param {Item} item The item changed.
 */
function sortItemAfterDelay(item) {
    bookmarkSorter.sortFolders(item.getFolder(), prefs.delay * SECOND);
}

/**
 * Sort `item` parent folder after the folder delay.
 * @param {Item} item The item changed.
 */
function sortItemAfterFolderDelay(item) {
    bookmarkSorter.sortFolders(item.getFolder(), prefs.folder_delay * SECOND);
}

/**
 * Sort recursively `item` folder after the preferred delay.
 * @param {Item} item The item changed.
 */
function sortChildrenAfterDelay(item) {
    function* getItemPlusChildrenFolders() {
        yield item;
        yield* item.getFolders();
    }

    bookmarkSorter.sortFolders(getItemPlusChildrenFolders(), prefs.delay * SECOND);
}

/**
 * Sort the item with the folder delay if it is a folder; otherwise, sort it with the normal delay.
 * @param {Item} item The item to sort.
 */
function sortFolderOrItem(item) {
    if(item instanceof Folder) {
        sortItemAfterFolderDelay(item);
    }
    else {
        sortItemAfterDelay(item);
    }
}

/**
 * On item change callback.
 */
function onChange(item, deleted, newFolder, annotationChange) {
    if(!deleted && !newFolder) {
        if(annotationChange) {
            sortChildrenAfterDelay(item);
        }
        else {
            sortFolderOrItem(item);
        }
    }
}

/**
 * On end batch callback.
 */
function onEndBatch() {
    bookmarkSorter.start();
}

/**
 * On item remove callback.
 */
function onRemove(item) {
    if(item instanceof Separator) {
        sortItemAfterDelay(item);
    }
}

/**
 * Add the bookmark observer.
 */
function addBookmarkObserver() {
    bookmarkManager.on('begin-batch', onBeginBatch);
    bookmarkManager.on('end-batch', onEndBatch);
    bookmarkManager.on('add', sortFolderOrItem);
    bookmarkManager.on('change', onChange);
    bookmarkManager.on('move', sortFolderOrItem);
    bookmarkManager.on('remove', onRemove);
    bookmarkManager.on('visit', sortItemAfterDelay);
}

/**
 * Sort all bookmarks.
 */
function sortAllBookmarks() {
    bookmarkSorter.sortAllBookmarks();
}

/**
 * Remove the bookmark observer.
 */
function removeBookmarkObserver() {
    bookmarkManager.removeListener('begin-batch', onBeginBatch);
    bookmarkManager.removeListener('end-batch', onEndBatch);
    bookmarkManager.removeListener('add', sortFolderOrItem);
    bookmarkManager.removeListener('change', onChange);
    bookmarkManager.removeListener('move', sortFolderOrItem);
    bookmarkManager.removeListener('remove', onRemove);
    bookmarkManager.removeListener('visit', sortItemAfterDelay);
}

/**
 * Adjust the auto sorting feature.
 */
function adjustAutoSort() {
    removeBookmarkObserver();
    if(prefs.auto_sort) {
        addBookmarkObserver();
        sortAllBookmarks();
    }
}

/**
 * Adjust the first run preference.
 */
function adjustFirstRun() {
    let prefName = getOptionName('firstrun');

    if(has(prefName)) {
        set(prefName, false);
    }
    else {
        set(prefName, true);
        prefs.auto_sort = false;
    }
}

/**
 * Sort if the auto sort option is on.
 */
function sortIfAuto() {
    if(prefs.auto_sort) {
        sortAllBookmarks();
    }
}

/**
 * Adjust the sort criteria of the bookmark sorter.
 */
function adjustSortCriteria() {
    let differentFolderOrder = prefs.folder_sort_order !== prefs.livemark_sort_order && prefs.folder_sort_order !== prefs.smart_bookmark_sort_order && prefs.folder_sort_order !== prefs.bookmark_sort_order;
    bookmarkSorter.setCriteria(sortCriterias[prefs.sort_by], prefs.inverse,
            sortCriterias[prefs.then_sort_by] || undefined, prefs.then_inverse,
            sortCriterias[prefs.folder_sort_by], prefs.folder_inverse,
            differentFolderOrder, prefs.case_insensitive
    );
    sortIfAuto();
}

/**
 * Create the events.
 */
function createEvents() {
    simplePrefs.on('auto_sort', function() {
        adjustAutoSort();
    });

    let preferences = ['folder_sort_order', 'livemark_sort_order', 'smart_bookmark_sort_order', 'bookmark_sort_order'];

    for(let preference of preferences) {
        simplePrefs.on(preference, sortIfAuto);
    }

    simplePrefs.on('case_insensitive', adjustSortCriteria);
    simplePrefs.on('sort_by', adjustSortCriteria);
    simplePrefs.on('then_sort_by', adjustSortCriteria);
    simplePrefs.on('folder_sort_by', adjustSortCriteria);
    simplePrefs.on('inverse', adjustSortCriteria);
    simplePrefs.on('then_inverse', adjustSortCriteria);
    simplePrefs.on('folder_inverse', adjustSortCriteria);
    simplePrefs.on('folder_sort_order', adjustSortCriteria);
    simplePrefs.on('livemark_sort_order', adjustSortCriteria);
    simplePrefs.on('smart_bookmark_sort_order', adjustSortCriteria);
    simplePrefs.on('bookmark_sort_order', adjustSortCriteria);

    simplePrefs.on('exclude_folders', showConfigureFoldersToExclude(sortIfAuto));
}

/**
 * Show the addon options.
 */
function showOptions() {
    windowUtils.getMostRecentBrowserWindow().BrowserOpenAddonsMgr('addons://detail/' + self.id + '/preferences');
}

/**
 * Show confirmation on install.
 */
function showConfirmation() {
    tabs.open({
        url: data.url('confirmation.html'),
        onOpen: function(tab) {
            tab.on('ready', function() {
                let worker = tab.attach({
                    contentScriptFile: data.url('confirmation.js')
                });
                worker.port.on('button-clicked', function(message) {
                    switch(message) {
                        case 'button-yes':
                            prefs.auto_sort = true;
                            break;
                        case 'button-no':
                            // Do nothing.
                            break;
                        case 'button-change-options':
                            showOptions();
                            break;
                    }
                    tab.close();
                });

                worker.port.emit('show', data.url('icon-48.png'));
            });
        },
    });
}

/**
 * Create the widgets.
 */
function createWidgets(install) {
    ActionButton({
        id: 'auto-sort-bookmarks-sort-all-toolbar',
        icon: {
            16: './icon-16.png',
            32: './icon-32.png',
            64: './icon-64.png',
        },
        label: _('sort_bookmarks'),
        onClick: sortAllBookmarks,
    });

    if(install) {
        showConfirmation();
    }
}

/**
 * Unload function.
 */
function exit(reason) {
    if(reason === 'disable') {
        reset(getOptionName('firstrun'));
    }
}

/**
 * Migrate to set the default folder sort order and the default folders to exclude.
 */
function migrate(upgrade) {
    if(upgrade) {
        if(!prefs.migration) {
            prefs.migration = '2.7';
        }

        if(prefs.migration < '2.8') {
            if([0, 2, 4, 5].indexOf(prefs.sort_by) >= 0) {
                prefs.folder_sort_by = prefs.sort_by;
                prefs.folder_inverse = prefs.inverse;
            }

            if(prefs.sort_menu === false) {
                setDoNotSortAnnotation(menuFolder.id);
                setRecursiveAnnotation(menuFolder.id);
            }

            if(prefs.sort_toolbar === false) {
                setDoNotSortAnnotation(toolbarFolder.id);
                setRecursiveAnnotation(toolbarFolder.id);
            }

            if(prefs.sort_unsorted === false) {
                setDoNotSortAnnotation(unsortedFolder.id);
                setRecursiveAnnotation(unsortedFolder.id);
            }

            delete prefs.sort_menu;
            delete prefs.sort_toolbar;
            delete prefs.sort_unsorted;
        }

        prefs.migration = self.version;
    }
}

/**
 * Set the minimum and maximum of integer preferences.
 */
function setPreferenceMinimumMaximum() {
    let preferences = ['folder_sort_order', 'livemark_sort_order', 'smart_bookmark_sort_order', 'bookmark_sort_order'];
    for(let preference of preferences) {
        setPreferenceMinimum(preference, 1);
        setPreferenceMaximum(preference, 4);
    }

    setPreferenceMinimum('folder_delay', 3);
}

/**
 * Auto-sort Bookmarks is a firefox extension that automatically sorts bookmarks.
 */
function main(options) {
    migrate(options.loadReason === 'upgrade');
    adjustFirstRun();
    createWidgets(options.loadReason === 'install');
    adjustSortCriteria();
    adjustAutoSort();
    createEvents();
    setPreferenceMinimumMaximum();
    adjustFirstRun();
}

exports.main = main;
exports.onUnload = exit;

exports.adjustAutoSort = adjustAutoSort;
exports.adjustSortCriteria = adjustSortCriteria;
exports.createEvents = createEvents;
exports.setPreferenceMinimumMaximum = setPreferenceMinimumMaximum;
